### PR TITLE
add support for AMD's monitorx and mwaitx opcodes

### PR DESCRIPTION
**Pull request by kmar**
please see https://github.com/DynamoRIO/dynamorio/issues/4011

### CODE REVIEW

**Comment by derekbruening** [comment]
run arm tests

File: core/arch/x86/decode_table.c [hunk 1]
```diff
@@ -6559,7 +6563,8 @@ const instr_info_t rm_extensions[][8] = {
   { /* rm extension 2 */
     {OP_swapgs, 0xf80f0177, "swapgs", xx, xx, xx, xx, xx, mrm|o64, x, END_LIST},
     {OP_rdtscp, 0xf90f0177, "rdtscp", edx, eax, xx, xx, xx, mrm|xop, x, exop[10]},/*AMD-only*/
-    {INVALID,   0x0f0131, "(bad)", xx, xx, xx, xx, xx, no, x, NA},
+    {OP_monitorx, 0xfa0f0177, "monitorx",  xx, xx, eax, ecx, edx, mrm, x, END_LIST},/*AMD-only*/
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: So the address should be `rax` for 64-bit, and it varies by address prefix: so `s/eax/axAX/`.  It looks like `monitor` has the same problem: if you could do the same substitution there (or if you prefer to leave that to a separate PR I could see it either way).

Comment (derekbruening) [inline core/arch/x86/decode_table.c]: The other thing is whether the address should be treated as a memory reference or not but let's make that completely orthogonal to this PR.  I filed#4013.  Maybe you could add a comment here like `XXX i#4013: Treat address in xax as IR memref?`.

File: suite/tests/api/ir_x86_0args.h [hunk 2]
```diff
@@ -73,6 +73,8 @@ OPCODE(fxam, fxam, fxam, 0)
 OPCODE(sahf, sahf, sahf, 0)
 OPCODE(mwait, mwait, mwait, 0)
 OPCODE(monitor, monitor, monitor, 0)
+OPCODE(mwaitx, mwaitx, mwaitx, 0)
+OPCODE(monitorx, monitorx, monitorx, 0)
```
Comment (derekbruening) [inline suite/tests/api/ir_x86_0args.h]: Please also enable the binutils tests for these, which are currently disabled:

```
$ git grep -E 'mwaitx|monitorx'
third_party/binutils/test_decenc/test_decenc_x86.asm:        /* FIXME i3577: monitorx, AMD only. */
third_party/binutils/test_decenc/test_decenc_x86.asm:        /* FIXME i3577: mwaitx, AMD only. */
third_party/binutils/test_decenc/test_decenc_x86_64.asm:        /* FIXME i3577: monitorx, AMD only. */
third_party/binutils/test_decenc/test_decenc_x86_64.asm:        /* FIXME i3577: mwaitx, AMD only. */
```

**Comment by kmar** [comment]
ok, I'm having a problem now. I tried to modify the expected disassembly output, but it still doesn't match.
I can't run the test locally and the test output doesn't show a diff so I'm stuck at the moment

**Comment by kmar** [comment]
still out of luck, I'm afraid I'll need a some help with this one
I don't understand why it fails, because monitor uses just monitor and addr(16/32) monitor, but it doesn't seems to be the case with monitorx...

**Comment by kmar** [comment]
ok, so the diff showed that adding new instruction bytes shifted ALL the near jumps and calls, so I'd have to copy and commit full output files (which I don't have) to make this PR pass, resulting in hundreds of lines changed, jesus...

I'm sorry but I just ran out of patience, if you want this PR then you'll have to fix it yourself

**Comment by kmar** [comment]
done, relevant tests pass now, I moved monitorx/mwaitx to the end of the test file so that they don't interfere with rel32 disp for jxx/call instructions, keeping diff after match failure, proven to be extremely useful

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by AssadHashmi** [comment]
User @kmar added to Jenkins ACL for AArch64 precommit tests.

File: core/arch/x86/decode_table.c (hunk 1 above)
Comment (kmar) [reply core/arch/x86/decode_table.c]: done

Comment (kmar) [reply core/arch/x86/decode_table.c]: done, added comments above monitor and monitorx

File: suite/tests/api/ir_x86_0args.h (hunk 2 above)
Comment (kmar) [reply suite/tests/api/ir_x86_0args.h]: this one was very tricky and I had to reorder tests in binutils due to jxx/call + rel32; done though

File: suite/tests/runcmp.cmake [hunk 3]
```diff
@@ -68,5 +68,15 @@ if (NOT "${cmd_out}" STREQUAL "${str}")
   file(WRITE "${tmp}" "${cmd_out}")
   set(tmp2 "${cmp}-expect")
   file(WRITE "${tmp2}" "${str}")
+
+  set(DiffCMD "diff")
```
Comment (derekbruening) [inline suite/tests/runcmp.cmake]: style nit: no CamelCase

File: third_party/binutils/test_decenc/test_decenc_x86.asm [hunk 4]
```diff
@@ -311,13 +311,6 @@ GLOBAL_LABEL(FUNCNAME:)
         /* xsaves, CPL 0 instruction. */
         /* RAW(0f) RAW(c7) RAW(29) */
         RAW(66) RAW(0f) RAW(ae) RAW(39)
-        /* FIXME i3577: monitorx, AMD only. */
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/test_decenc_x86.asm]: These are grouped by the binutil test file name: see the comment at line 299.  They should be kept together. 
 Please move the whole group for that file as one.

File: third_party/binutils/test_decenc/test_decenc_x86_64.asm [hunk 5]
```diff
@@ -106433,5 +106426,13 @@ GLOBAL_LABEL(FUNCNAME:)
 
 #endif /* DISABLED_UNTIL_BUG_3581_IS_FIXED */
 
+        /* monitorx, AMD only. */
```
Comment (derekbruening) [inline third_party/binutils/test_decenc/test_decenc_x86_64.asm]: Ditto here: these are part of x86_64_arch_3.s.

File: suite/tests/api/ir_x86_0args.h (hunk 2 above)
Comment (derekbruening) [reply suite/tests/api/ir_x86_0args.h]: That is the downside of golden output tests, though making all these binutils tests into internal separate checks would be even more work.  I recall having to take the new full output on an update and verifying the final diff was all rel jmps: not too bad.

File: third_party/binutils/test_decenc/test_decenc_x86_64.asm (hunk 5 above)
Comment (kmar) [reply third_party/binutils/test_decenc/test_decenc_x86_64.asm]: have it your way (you apparently didn't read my comments), but the PR will grow from 54 to 1k lines changed

Comment (derekbruening) [reply third_party/binutils/test_decenc/test_decenc_x86_64.asm]: I would prefer to not move anything and to keep these lines where they were.  But if something is going to be moved then the group needs to be moved: splitting it up will cause confusion with the original binutils.

Comment (kmar) [reply third_party/binutils/test_decenc/test_decenc_x86_64.asm]: that's what I did, I went back and kept everything inplace as it originally was, grouped, nothing moved, but the diff for expected results will be quite large. tests pass so I'm hoping for a final review for this PR

Comment (kmar) [reply third_party/binutils/test_decenc/test_decenc_x86_64.asm]: a side note:
whoever commented out the unsupported instructions forgot to replace them with nops.

the problem is that each relative call is disassembled with absolute taget address, so
e8 00 00 00 00  => call $...absolute_address
depends on number of bytes before it, this is the problem I was talking about

best would be to either force the disassembler to use relative target, like call $+0 (if that's possible at all) or fix the unsupported instructions with nop-padding

otherwise anyone who's adding tests for new instructions will have 1k-line changes in xxx-expect

but this outside the scope of this PR

Comment (derekbruening) [reply third_party/binutils/test_decenc/test_decenc_x86_64.asm]: Yes, I understand and commented earlier on having hit it before and tradeoffs of this "golden output" approach.  Putting in nops is a good idea.  At this point we are willing to live w/ large-ish diffs in the expect file; could certainly be improved.

**Comment by derekbruening** [comment]
> umm, I certainly didn't start a review of my own changes. what did I mess up?
> any way to get rid of this?

I think I saw that too in a recent PR and am not sure how it happened: presumably some Github bug.  Just ignore it.
